# 4.04 Continued Merges
### 01. Import orders_products_combined.pkl
### 02. Check the shape of df_merged_large
### 03. Combine df_merged_large with df_prods
### 04. Confirm the results of the merge
### 05. Export

## 01. Imports

In [1]:
#Import libraries - pandas, Numpy, OS
import pandas as pd
import numpy as np
import os

In [2]:
#Creating a path to the root directory for this project
path = r'C:\Users\theva\OneDrive\CareerFoundry\Data Immersion\A4\Instacart Basket Analysis 11-2022'

In [3]:
#Importing products.csv to df_prods
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

In [4]:
#Importing orders_products_combined.pkl to df_merged_large
df_merged_large = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.pkl'))

## 02. Check the shape of df_merged_large

In [5]:
#Setting verification list variable
orig_shape = 32434489, 10

In [6]:
#Setting new shape to list variable
new_shape = df_merged_large.shape
new_shape

(32434489, 10)

In [7]:
#Testing the shape
orig_shape == new_shape

True

## 03. Combine df_merged_large with df_prods

In [8]:
#Dropping previous merge flag
df_merged_large = df_merged_large.drop(columns = '_merge')

In [9]:
#Checking shape
df_merged_large.shape

(32434489, 9)

In [10]:
#Looking for best merge key
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [11]:
#Looking for best merge key
df_merged_large.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  
0                   999.0         196                  1          0  
1                   999.0       14084                  2          0  
2                   999.0       12427                  3          0  
3                   999.0       26088                  4          0  
4                   999.0       26405                  5          0

Merge key will be 'product_id'

In [14]:
#Merging df_merged_large with df_prods
df_ords_prods_merged = df_merged_large.merge(df_prods, on = 'product_id', indicator = True)

In [39]:
#Checking merged DF content
df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                   999.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge  
0         Soda        77              7     9.0   both  
1         Soda        77              7     9.0   both  
2         Soda        77              7     9.0   both  
3         Soda        77              7     9.0   both  
4         Soda        77              7     9.0   both

## 04. Confirm the results of the merge

In [41]:
#Checking merge flag frequency
df_ords_prods_merged['_merge'].value_counts()

both          32404859
left_only            0
right_only           0
Name: _merge, dtype: int64

There are only 'both' flags since this was an inner join. Next we can compare the shapes of the current and previous versions to see how the rows and columns changed.

In [37]:
#Checking new shape
new_shape = pd.DataFrame(df_ords_prods_merged.shape)

In [38]:
#Checking change of shape
orig_shape = pd.DataFrame(df_merged_large.shape)
new_shape - orig_shape

0
0 -29630
1      5

The merge added 5 new columns (4 columns from df_prods and 1 merge flag) and lost 29,630 rows where the two data frames had no converging product keys.

## 05. Export

In [42]:
#Dropping the merge flag
df_ords_prods_merged = df_ords_prods_merged.drop(columns = '_merge')

In [44]:
#Verifying shape before export
df_ords_prods_merged.shape

(32404859, 13)

In [45]:
#Exporting merged dataframe
df_ords_prods_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged.pkl'))